### DISCUSSION TITLE
前缀和&差分&倍增

### DISCUSSION THREAD

**Discussion by JYZ-XiaoNB** [CSGrandeur/s-1problem1day1ac#433] [Algorithm]
### 前缀和算法
前缀和的最主要目的就是求子数组的和的大小。我们维护一个由前缀的和组成的数组sum，则求a[1]到a[3]的和：a[1] + a[2] + a[3] = sum[3] - sum[0]
### 例题：子段求和
_给出一个长度为N的数组，进行Q次查询，查询从第i个元素开始长度为l的子段所有元素之和。
例如，1 3 7 9 -1，查询第2个元素开始长度为3的子段和，1 {3 7 9} -1。3 + 7 + 9 = 19，输出19。_
### 代码实现
```C++
#include<iostream>
#include<cstring>
#include<algorithm>
#include<cstdio>
#include<map>
#include<set> 
#include<string>
#include<stack>
using namespace std;
int main()
{
	long long int n,a,s[50001];
	cin>>n;
	for(int i=1;i<=n;i++)
	{
		cin>>a;
		s[i]=s[i-1]+a;
	}
	int q,l,i;
	cin>>q;
	while(q--)
	{
		cin>>i>>l;
		cout<<s[l+i-1]-s[i-1]<<endl;
	} 
	return 0;
}
```
### 二维前缀和
- 原二维数组{a[1][1], a[1][2], a[1][3], …, a[n][m]}，注意：数组下标从1开始，同时令S[0][0] = 0。
- 前缀和S[i][j]：S[i][j] = a[1][1] + a[1][2] + … + a[i][j]
- **求S[i][j]**       for循环遍历一下原二维数组，其中S[i][j] = S[i-1][j] + S[i][j-1] - S[i-1][j-1] + a[i][j]

![aHR0cHM6Ly91cGxvYWQtaW1hZ2VzLmppYW5zaHUuaW8vdXBsb2FkX2ltYWdlcy82ODU3NzY0LTEwMTJiYWExMDAxODEyYmYucG5n](https://user-images.githubusercontent.com/93431491/162111082-014a150a-a30e-4424-9009-c80359612a1c.png)

- **求任意一个子矩阵的和，如{[x1][y1],[x2][y2]}的和**    S[x2][y2] - S[x2][y1-1] - S[x1-1][y2] + S[x1-1][y1-1]
### 例题：最大子矩阵
_给你一个m×n的整数矩阵，在上面找一个x×y的子矩阵，使子矩阵中所有元素的和最大。_
```C++
#include<iostream>
#include<string.h>
#include<algorithm>
using namespace std;
const int Len = 1005;
int map[Len][Len];   
long long S[Len][Len];      
int main()
{
    int t;
    scanf("%d", &t);
    while(t --)
    {
        memset(S , 0 , sizeof(S));
        int m,n,x,y;
        scanf("%d %d %d %d", &m, &n, &x, &y); //读入矩阵大小，和需要求的子矩阵大小X*Y
        for(int i = 1; i <= m; i ++)
            for(int j = 1; j <= n; j ++)
            {
                scanf("%d", &map[i][j]);
                S[i][j] = S[i - 1][j] + S[i][j - 1] - S[i - 1][j - 1] + map[i][j];//求S[i][j]
            }
        long long ans = -1;
        for(int i = 1; i <= m - x + 1; i ++)
            for(int j = 1; j <= n - y + 1; j ++)
                ans = max(ans , S[i + x - 1][j + y - 1] - S[i + x - 1][j - 1] - S[i - 1][j + y - 1] + S[i - 1][j - 1]);
                                    //max函数的第二个参数就是求给定长宽后的矩形区域的数字和
        printf("%lld\n",ans);
    }
    return 0;
}
```
### 差分
什么是差分？差分是一个数组相邻两元素的差，一般为下标靠后的减去靠前的一个。设差分数组p[]，即：**p[i]=a[i]-a[i-1]**
### 差分和前缀和的联系
类似于数学中的求导和积分，差分可以看成前缀和的逆运算。

首先给定一个原数组a：a[1], a[2], a[3],,,,,, a[n];

然后我们构造一个数组b ： b[1] ,b[2] , b[3],,,,,, b[i];

使得 a[i] = b[1] + b[2 ]+ b[3] +,,,,,, + b[i]
 
也就是说，a数组是b数组的前缀和数组，反过来我们把b数组叫做a数组的差分数组。换句话说，每一个a[i]都是b数组
中从头开始的一段区间和。
### 差分的作用
**举个栗子：**
给定区间[l ,r ]，让我们把a数组中的[ l, r]区间中的每一个数都加上c,即 a[l] + c , a[l+1] + c , a[l+2] + c ,,,,,, a[r] + c;

暴力做法是for循环l到r区间，时间复杂度O(n)，如果我们需要对原数组执行m次这样的操作，时间复杂度就会变成O(n*m)。
**使用差分：**
始终要记得，a数组是b数组的前缀和数组，比如对b数组的b[i]的修改，会影响到a数组中从a[i]及往后的每一个数。

首先让差分b数组中的 b[l] + c ,通过前缀和运算，a数组变成 a[l] + c ,a[l+1] + c,,,,,, a[n] + c;

然后我们打个补丁，b[r+1] - c, 通过前缀和运算，a数组变成 a[r+1] - c,a[r+2] - c,,,,,,,a[n] - c;
_因为通过b[l]+c,会使a数组在l之后的元素都+c，而我们只需要在区间[l,r]中操作，所以需要b[r+1]-c_

因此我们得出一维差分结论：给a数组中的[ l, r]区间中的每一个数都加上c,只需对差分数组b做 b[l] + = c, b[r+1] - = c。时间复杂度为O(1), 大大提高了效率。
### 二维差分
a[][]数组是b[][]数组的前缀和数组，那么b[][]是a[][]的差分数组

原数组： a[i][j]

我们去构造差分数组： b[i][j]

使得a数组中a[i][j]是b数组左上角(1,1)到右下角(i,j)所包围矩形元素的和
### b数组二维差分操作
**首先学习如何让a数组的子矩阵为 以(x1,y1)为左上角，以(x2,y2)为右下角所围成的矩形区域内的数+c：**
b[x1][y1] + = c;

b[x1,][y2+1] - = c;

b[x2+1][y1] - = c;

b[x2+1][y2+1] + = c;
![20201217171134926](https://user-images.githubusercontent.com/93431491/162376203-371ca725-5566-4b7d-a300-d9920fb9a776.png)
_b[x1][ y1 ] +=c ; 对应图1 ,让整个a数组中蓝色矩形面积的元素都加上了c。
b[x1,][y2+1]-=c ; 对应图2 ,让整个a数组中绿色矩形面积的元素再减去c，使其内元素不发生改变。
b[x2+1][y1]- =c ; 对应图3 ,让整个a数组中紫色矩形面积的元素再减去c，使其内元素不发生改变。
b[x2+1][y2+1]+=c; 对应图4,让整个a数组中红色矩形面积的元素再加上c，红色内的相当于被减了两次，再加上一次c，才能使其恢复_
**将这个操作写为函数**
```C++
void insert(int x1,int y1,int x2,int y2,int c)
{ 
    b[x1][y1]+=c;
    b[x2+1][y1]-=c;
    b[x1][y2+1]-=c;
    b[x2+1][y2+1]+=c;
}
```
**再学习如何构造b数组**
我们可以先假想a数组为空，那么b数组一开始也为空，但是实际上a数组并不为空，因此我们每次让以(i,j)为左上角到以(i,j)为右下角面积内元素(其实就是一个小方格的面积)去插入 c=a[i][j]，等价于原数组a中(i,j) 到(i,j)范围内 加上了 a[i][j] ,因此执行n*m次插入操作，就成功构建了差分b数组.
```C++
  for(int i=1;i<=n;i++)
  {
      for(int j=1;j<=m;j++)
      {
          insert(i,j,i,j,a[i][j]);   
      }
  }
```
**总结简化为公式：**         b[i][j]=a[i][j]−a[i−1][j]−a[i][j−1]+a[i−1][j−1]
### 倍增
从字面的上意思看就是成倍的增长 ,这是指我们在进行递推时,如果状态空间很大,通常的线性递推无法满足时间和空间复杂度的要求 ,那么我们就可以通过成倍的增长,只递推状态空间中在 2 的整数次幂位置上的值作为代表 。当需要其他位置上的值时,我们只需要通过" 任意整数可以表示成若干个2的次幂项的和 " 这一性质(13= 2^3 + 2^2 +2^0），使用之前求出的代表值拼成所需的值。
![1352815470_5150](https://user-images.githubusercontent.com/93431491/162379238-7a3acce9-8e21-4e32-81c7-2fbe270cf121.jpg)
_从A出发跳1步到1（记录下来）
从1出发跳1步到2（记录下来）
…………（跳1步的记录完毕）
从A出发跳2步？就是从A出发跳1步再跳1步的到的地方,直接记录从1出发跳1步会到的2这个格子作为A跳2步会到的格子
从1出发跳2步？跟刚才一样，直接记录。
…………（跳2步的记录完毕）_
**在记录A跳n步会到的格子的时候，就可以直接从A跳n步看到了哪里再记录，跳n步跟跳1步的代价是一样的**
### 举个栗子
一个数组a {2,5,7,11,19} 中查找最大的小于12的数字k。
常规做法：先预处理前缀和 , (a [ i ] > 0 )然后用 二分找到 这个k , 这个复杂度最坏是O ( n ) ,因为如果我们的 询问的T 太小 ,就还不如从前往后枚举。
我们可以设计这样一个 倍增算法 : 设定一个增长长度 p 和已确定长度 l，现在要确定 a[l+p] 是否满足条件，若满足条件（比12小），则 p 成2倍增长；否则 p 缩小范围（试着缩小范围判断条件）
```C++
l=0;
p=1;
while(p)//如果还能扩增范围(l)就继续
{
    if(a[l+p]<12)
    {
        l+=p;//增加已知范围
        p*=2//成倍增长
    }
    else
    {
        p=p/2;//缩小范围
    }
}
cout<<a[l];

```
### 例题：求解 RMQ 问题
给出 n 个数，有 m 次询问，每次输出区间 [x,y] 的最大值。
**结合 DP，设 f[i][j] 表示从 i 开始的长度为 2j 的区间的最大值，那么容易得出方程：f[i][j]=max(f[i][j-1],f[i+2(j-1)][j-1])。其中 f[i][j-1] 表示区间 [i,i+2(j-1)-1] 的最大值，f[i+2(j-1)][j-1] 表示区间 [i+2(j-1),i+2j]的最大值。

现在要处理询问了。从 x,y 中我们可以找到一个值 k，使得 [x,x+2k-1] 和 [y-2k+1,y]可以正好覆盖 [x,y] 整个区间，那么 [x,y] 的区间最大值就是 max(f[x][k],f[y-2k,k])。cmath库的函数 log2 可以帮我们解决求 k 的问题。**
```C++
#include <iostream> 
#include <cstdio>
#include <cmath>
#include <string>
#include <sstream>
#include <algorithm>
using namespace std;
int main()
{
	int m,n,x,y;
	int a[200];
	int f[300][300];
	cin>>n>>m;
	for(int i=1; i<=n; i++)
	{
		cin>>a[i];
	}
	for(int i=1;i<=n;i++)
		f[i][0]=a[i];//一个数的区间就是其本身
	for(int j=1;(1<<j)<=n;j++)
    	for(int i=1;i+(1<<j)-1<=n;i++)
        	f[i][j]=max(f[i][j-1],f[i+(1<<j)][j-1]);//预处理f数组 
	for(int i=1;i<=m;i++)
	{
    	cin>>x>>y;
    	int k=log2(y-x+1);
    	cout<<max(f[x][k],f[y-(1<<k)+1][k])<<endl;
	} 
return 0;
}
```
### 参考资料
[CSDN倍增算法基础](https://blog.csdn.net/m0_57344422/article/details/118080399?ops_request_misc=&request_id=&biz_id=102&utm_term=%E5%80%8D%E5%A2%9E&utm_medium=distribute.pc_search_result.none-task-blog-2~all~sobaiduweb~default-0-118080399.142^v7^pc_search_result_control_group,157^v4^control&spm=1018.2226.3001.4187)
[CSDN前缀和与差分](https://blog.csdn.net/weixin_45629285/article/details/111146240?ops_request_misc=%257B%2522request%255Fid%2522%253A%2522164930241016782184627467%2522%252C%2522scm%2522%253A%252220140713.130102334.pc%255Fall.%2522%257D&request_id=164930241016782184627467&biz_id=0&utm_medium=distribute.pc_search_result.none-task-blog-2~all~first_rank_ecpm_v1~rank_v31_ecpm-1-111146240.142^v5^pc_search_result_cache,157^v4^control&utm_term=%E5%89%8D%E7%BC%80%E5%92%8C%E5%92%8C%E5%B7%AE%E5%88%86&spm=1018.2226.3001.4187)
[前缀和算法原理及代码](https://blog.csdn.net/cdlwhm1217096231/article/details/109609423?ops_request_misc=%257B%2522request%255Fid%2522%253A%2522164929423016780271965893%2522%252C%2522scm%2522%253A%252220140713.130102334.pc%255Fall.%2522%257D&request_id=164929423016780271965893&biz_id=0&utm_medium=distribute.pc_search_result.none-task-blog-2~all~first_rank_ecpm_v1~rank_v31_ecpm-7-109609423.142^v5^pc_search_result_cache,157^v4^control&utm_term=%E4%BA%8C%E7%BB%B4%E5%89%8D%E7%BC%80%E5%92%8C&spm=1018.2226.3001.4187)
[【白话系列】倍增算法](https://blog.csdn.net/JarjingX/article/details/8180560?ops_request_misc=%257B%2522request%255Fid%2522%253A%2522164939947616780264035461%2522%252C%2522scm%2522%253A%252220140713.130102334..%2522%257D&request_id=164939947616780264035461&biz_id=0&utm_medium=distribute.pc_search_result.none-task-blog-2~all~top_positive~default-1-8180560.142^v7^pc_search_result_control_group,157^v4^control&utm_term=%E5%80%8D%E5%A2%9E&spm=1018.2226.3001.4187)
### 参考题单
<html>
<body>

序号 | 题号 | 标题 | 题型 | 难度评级
-- | -- | -- | -- | --
1 | 洛谷P5495 | Dirichlet 前缀和 | 一维前缀和 | ⭐
2 | 洛谷P5488 | 差分与前缀和 | 一维前缀和+差分 | ⭐
3 | 洛谷P4515 | 上帝造题的七分钟 | 二维前缀和+差分 | ⭐⭐
4 | 牛客网 | 减成一 | 差分 | ⭐
5 | Codeforces 1000C | Covered Points Count | 离散化+前缀和 | ⭐⭐
6 | 洛谷P1438 | 无聊的数列 | 递归+差分 | ⭐⭐⭐
7 | 洛谷P3902 | 递增 | dp+倍增 | ⭐⭐⭐
8 | 洛谷P1306 | 斐波那契公约数 | 递归+倍增 | ⭐⭐⭐
9 | 洛谷P5353 | 树上后缀排序 | 后缀数+倍增 | ⭐⭐⭐⭐
10 | 洛谷P4155 | 国旗计划 | 贪心+倍增 | ⭐⭐⭐
11 | HDU P3712 | Fiolki | 并查集+倍增 | ⭐⭐⭐⭐⭐

</body>
</html>

**Comment by re-resolve** [comment]
最后的例题，有些难理解（😵）
